lib/active_merchant/billing/gateways/pro_pay.rb

```diff
@@ -0,0 +1,327 @@
+        request = build_xml_request do |xml|
+          add_invoice(xml, money, options)
+          add_payment(xml, payment, options)
+          add_address(xml, options)
+          add_account(xml, options)
+          add_recurring(xml, options)
+          xml.transType "05"
+        end
+
+        commit(request)
+      end
+
+      def capture(money, authorization, options={})
+        request = build_xml_request do |xml|
+          xml.amount money if money
+          xml.transNum authorization
+          add_account(xml, options)
+          xml.transType "06"
+        end
+
+        commit(request)
+      end
+
+      def refund(money, authorization, options={})
+        request = build_xml_request do |xml|
+          xml.transNum authorization
+          xml.amount money if money
+          xml.currencyCode if options[:currency]
+          add_account(xml, options)
+          xml.transType "07"
+        end
+
+        commit(request)
+      end
+
+      def void(authorization, options={})
+        refund(nil, authorization, options)
+      end
+
+      def credit(money, payment, options={})
+        request = build_xml_request do |xml|
+          add_invoice(xml, money, options)
+          add_payment(xml, payment, options)
+          add_account(xml, options)
+          xml.transType "35"
+        end
+
+        commit(request)
+      end
+
+      def verify(credit_card, options={})
+        MultiResponse.run(:use_first_response) do |r|
+          r.process { authorize(100, credit_card, options) }
+          r.process(:ignore_result) { void(r.authorization, options) }
+        end
+      end
+
+      def supports_scrubbing?
+        true
+      end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r((<certStr>).+(</certStr>)), '\1[FILTERED]\2').
+          gsub(%r((<ccNum>).+(</ccNum>)), '\1[FILTERED]\2').
+          gsub(%r((<CVV2>).+(</CVV2>)), '\1[FILTERED]\2')
+      end
+
+      private
+
+      def add_payment(xml, payment, options)
+        xml.ccNum payment.number
+        xml.expDate "#{format(payment.month, :two_digits)}#{format(payment.year, :two_digits)}"
+        xml.CVV2 payment.verification_value
+        xml.cardholderName payment.name
+      end
+
+      def add_address(xml, options)
+        if address = options[:billing_address] || options[:address]
+          xml.addr address[:address1]
+          xml.aptNum address[:address2]
+          xml.city address[:city]
+          xml.state address[:state]
+          xml.zip address[:zip]
+        end
+      end
+
+      def add_account(xml, options)
+        xml.accountNum options[:account_num]
+      end
+
+      def add_invoice(xml, money, options)
+        xml.amount amount(money)
+        xml.currencyCode options[:currency] || currency(money)
+        xml.invNum options[:order_id] || SecureRandom.hex(25)
+      end
+
+      def add_recurring(xml, options)
+        xml.recurringPayment options[:recurring_payment]
```

Comment:
I think that does make sense; I'll try to remember to follow that in the future.